test/test_overrides.py

```diff
@@ -0,0 +1,756 @@
+import torch
+import numpy as np
+import unittest
+import six
+import inspect
+import pprint
+
+from common_utils import TestCase
+
+HANDLED_FUNCTIONS = {}
+HANDLED_FUNCTIONS_SUB = {}
+HANDLED_FUNCTIONS_SUB_DIAGONAL = {}
+HANDLED_FUNCTIONS_TENSOR_LIKE = {}
+
+def implements(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS[torch_function.__name__] = func
+        return func
+    return decorator
+
+def implements_sub(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS_SUB[torch_function.__name__] = func
+        return func
+    return decorator
+
+def implements_sub_diagonal(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS_SUB_DIAGONAL[torch_function.__name__] = func
+        return func
+    return decorator
+
+def implements_tensor_like(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS_TENSOR_LIKE[torch_function.__name__] = func
+        return func
+    return decorator
+
+class ImplementationMeta(type):
```

Comment:
What does this do?

Reply:
Instead of adding explicit definitions for the overrides, a class can use this metaclass to define the overrides dynamically based off of the `TORCH_IMPLEMENTATIONS` and `IMPLEMENTS_DECORATOR` class attributes. It means adding an override gets reduced from 5-6 lines to 1 line, which we thought would be more maintainable, especially for the `TensorLike` tests.

In principle we could expose this metaclass for authors who want to use the `__torch_function__` protocol but for now this is just here to make the tests simpler.

Thanks for pointing out the docstrings are poor, I will update them.